File: cpp/0067-add-binary.cpp
#include <algorithm>
#include <string>

class Solution {
    public:
        [[nodiscard]] auto addBinary(const std::string &a, const std::string &b)
            -> std::string {
            auto ans = std::string{};
            auto i = a.size();
            auto j = b.size();
            auto carry = 0;

            while (i > 0 || j > 0 || carry > 0) {
                if (i > 0) {
                    carry += a[i - 1] - '0';
                    --i;
                }
                if (j > 0) {
                    carry += b[j - 1] - '0';
                    --j;
                }

                ans.push_back(static_cast<char>(carry % 2 + '0'));
                carry /= 2;
            }

            std::reverse(ans.begin(), ans.end());

            return ans;
        }
};

File: py/0067-add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        # return bin(int(a, 2) + int(b, 2))[2:]

        # binary result bits
        ans: list[str] = []
        # pointers
        i = len(a) - 1
        j = len(b) - 1
        carry = 0
        # loop while there are bits to process or a carry
        while i >= 0 or j >= 0 or carry > 0:
            if i >= 0:
                carry += int(a[i])
                i -= 1
            if j >= 0:
                carry += int(b[j])
                j -= 1

            # append current bit to result
            ans.append(str(carry % 2))
            # update carry
            carry //= 2

        # reverse list and convert to str
        return "".join(ans)[::-1]
